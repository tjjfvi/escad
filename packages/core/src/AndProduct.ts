import { Product, _Product, checkTypeProperty, ProductType } from "."

export interface AndProduct<T extends readonly Product[] = readonly Product[]> extends _Product {
  readonly type: "AndProduct",
  readonly children: T,
}

export const AndProduct = {
  create: <T extends readonly Product[]>(children: T): AndProduct<T> => ({
    type: "AndProduct",
    children,
  }),
  isAndProduct: checkTypeProperty.string<AndProduct>("AndProduct"),
  getAndProductType: <T extends AndProduct>(product: T): AndProductType<T> =>
    AndProductType.create(product.children.map(Product.getProductType)) as any,
}

export interface AndProductType<T extends AndProduct = AndProduct> {
  readonly type: "AndProductType",
  readonly elementTypes: (
    T extends AndProduct<infer U>
      ? { [K in keyof U]: U[K] extends Product ? ProductType<U[K]> : never }
      : never
  ),
}

export const AndProductType = {
  create: (elementTypes: ProductType[]): AndProductType => ({
    type: "AndProductType",
    elementTypes,
  }),
  isAndProductType: checkTypeProperty.string<AndProductType>("AndProductType"),
  getAndProductType: <T extends AndProduct>(product: T): AndProductType<T> =>
    AndProductType.create(product.children.map(x => Product.getProductType(x))) as any,
}
