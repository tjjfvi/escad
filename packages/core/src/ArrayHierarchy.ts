
import { checkTypeProperty } from "./checkTypeProperty";
import { _Hierarchy, Hierarchy } from "./Hierarchy";

export interface ArrayHierarchy extends _Hierarchy {
  readonly type: "ArrayHierarchy",
  readonly children: Hierarchy[],
}

export const ArrayHierarchy = {
  create: ({
    children,
    linkedProducts = children.flatMap(x => x.linkedProducts ?? []),
  }: Omit<ArrayHierarchy, "type">): ArrayHierarchy => ({
    type: "ArrayHierarchy",
    children,
    linkedProducts,
  }),
  from: (array: unknown[], raw = false) =>
    ArrayHierarchy.create({
      children: array.map(x => Hierarchy.from(x, raw))
    }),
  isArrayHierarchy: checkTypeProperty.string<ArrayHierarchy>("ArrayHierarchy"),
}
