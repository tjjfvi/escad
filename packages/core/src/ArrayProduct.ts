
import { checkTypeProperty } from "./checkTypeProperty";
import { Product, _Product, ProductType } from "./Product";

export interface ArrayProduct<T extends Product = Product> extends _Product {
  readonly type: "ArrayProduct",
  readonly children: readonly T[],
}

export const ArrayProduct = {
  create: <T extends Product>(children: readonly T[]): ArrayProduct<T> => {
    if(children.length === 0)
      throw new Error("ArrayProducts cannot be empty");
    return {
      type: "ArrayProduct",
      children,
    };
  },
  isArrayProduct: checkTypeProperty<ArrayProduct>("ArrayProduct"),
  getArrayProductType: <T extends ArrayProduct>(product: T): ArrayProductType<T> =>
    ArrayProductType.create(Product.getProductType(product.children[0])) as any
}

export interface ArrayProductType<T extends ArrayProduct = ArrayProduct> {
  readonly type: "ArrayProductType",
  readonly elementType: ProductType<T extends ArrayProduct<infer U> ? U : never>,
}

export const ArrayProductType = {
  create: (elementType: ProductType): ArrayProductType => ({
    type: "ArrayProductType",
    elementType,
  }),
  isArrayProductType: checkTypeProperty<ArrayProductType>("ArrayProductType"),
}
