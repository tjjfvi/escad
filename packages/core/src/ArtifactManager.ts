
import fs from "fs-extra";
import { Readable } from "stream";

import { Sha } from "./hash";
import { ReadonlyArtifactManager } from "./ReadonlyArtifactManager";

export abstract class ArtifactManager<T> extends ReadonlyArtifactManager<T> {

  abstract deserialize(stream: Readable): Promise<T | null>;

  async lookup(sha: Sha){
    return await this.cache.getAsync(sha.hex, async () => {
      const path = await this.getPath(sha);
      let stream = fs.createReadStream(path);
      return await new Promise(resolve => {
        stream
          .once("open", () => {
            resolve(this.deserialize(stream));
          })
          .once("error", () => resolve(null));
      })
    })
  }

  abstract getSha(t: T): Promise<Sha>

  reference = () => Sha.reference().map<T>({
    serialize: t => this.getSha(t),
    deserialize: async sha => {
      const artifact = await this.lookup(sha);
      if(!artifact)
        throw new Error(`Could not find artifact of sha ${sha}`);
      return artifact;
    }
  })

}
