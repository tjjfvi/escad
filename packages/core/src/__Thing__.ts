
import { ExtensibleFunction } from "./ExtensibleFunction";

export class __Thing__ extends ExtensibleFunction {

  declare protected __thing__: true;

}
