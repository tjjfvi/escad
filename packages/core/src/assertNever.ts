
import { inspect } from "util"

export function assertNever(
  value: never,
  message: (str: string) => string = str => `Expected never, got ${str}`,
): never{
  throw new Error(message(inspect(value)))
}
