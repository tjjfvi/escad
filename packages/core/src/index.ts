
export * from "./builtins";
import builtins from "./builtins";
import Element, { Elementish } from "./Element";
import Product from "./Product";

const escadFunc = <T extends Product>(...a: Elementish<T>[]) => {
  return new Element(a.length === 1 ? a[0] : a);
}

const escad = new Proxy(escadFunc, {
  get: (target, key) =>
    key in target ?
      target[key as keyof typeof target] :
      builtins[key as keyof typeof builtins]
}) as typeof escadFunc & typeof builtins;

export default escad;

export { default as Component } from './Component';
export { default as Element } from './Element';
export { default as Hierarchy } from './Hierarchy';
export { default as Id } from './Id';
export { default as Operation } from './Operation';
export { default as Product } from './Product';
export { default as Work } from './Work';
export { default as ProductManager } from './ProductManager';
export * from './Component';
export * from './Element';
export * from './Hierarchy';
export * from './Id';
export * from './Operation';
export * from './Product';
export * from './Work';
export * from "./builtins";
export * from "./mapOperation";
export * from "./ExportType";

