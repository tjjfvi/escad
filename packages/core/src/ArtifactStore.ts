
import { ArtifactManager } from "./ArtifactManager"
import { Hash } from "./Hash"
import { WrappedValue } from "./WrappedValue"

export interface ArtifactStore {
  readonly storeRaw?: (hash: Hash<unknown>, artifact: WrappedValue, artifactManager: ArtifactManager) => Promise<void>,
  readonly storeRef?: (loc: readonly unknown[], hash: Hash<unknown>, artifactManager: ArtifactManager) => Promise<void>,
  readonly lookupRaw?: (hash: Hash<unknown>, artifactManager: ArtifactManager) => Promise<WrappedValue | null>,
  readonly lookupRef?: (loc: readonly unknown[], artifactManager: ArtifactManager) => Promise<WrappedValue | null>,
}
