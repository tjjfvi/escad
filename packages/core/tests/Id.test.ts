import { Id } from "../src"

describe("Id.create", () => {
  test("@escad/core/Test/test0", () => {
    expect(Id.create(__filename, "@escad/core", "Test", "test0")).toMatchSnapshot()
  })
  test("Throws on duplicate", () => {
    expect(() => Id.create(__filename, "@escad/core", "Test", "test0")).toThrowErrorMatchingSnapshot()
  })
  test("Throws on incorrect package name", () => {
    expect(() => Id.create(__filename, "@escad/incorrect", "Test", "test0")).toThrowErrorMatchingSnapshot()
  })
})
