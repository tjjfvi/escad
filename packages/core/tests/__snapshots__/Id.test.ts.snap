// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Id.create @escad-core-0-test0 1`] = `
Object {
  "full": "@escad-core-0-test0",
  "isId": true,
  "name": "test0",
  "packageName": "@escad/core",
  "version": "0",
}
`;

exports[`Id.create Throws on duplicate 1`] = `"Duplicate ids created under @escad-core-0-test0"`;

exports[`Id.create Throws on incorrect package name 1`] = `"Id.create: packageName mismatch; @escad/core attempted to create an id under @escad/incorrect"`;

exports[`Id.create Throws on incorrect package version 1`] = `"Id.create: version mismatch; @escad/core@0.0.0 attempted to create an id under vIncorrect"`;
